crates/nu-command/src/filters/join.rs

```diff
@@ -0,0 +1,411 @@
+                        Value::Int {
+                            val: 3,
+                            span: Span::test_data(),
+                        },
+                    ],
+                    span: Span::test_data(),
+                }],
+                span: Span::test_data(),
+            }),
+        }]
+    }
+}
+
+fn join_type(call: &Call) -> Result<JoinType, nu_protocol::ShellError> {
+    match (
+        call.has_flag("inner"),
+        call.has_flag("left"),
+        call.has_flag("right"),
+        call.has_flag("outer"),
+    ) {
+        (_, false, false, false) => Ok(JoinType::Inner),
+        (false, true, false, false) => Ok(JoinType::Left),
+        (false, false, true, false) => Ok(JoinType::Right),
+        (false, false, false, true) => Ok(JoinType::Outer),
+        _ => Err(ShellError::UnsupportedInput(
+            "Choose one of: --inner, --left, --right, --outer".into(),
+            "".into(),
+            call.head,
+            call.head,
+        )),
+    }
+}
+
+fn join(
+    left: &Vec<Value>,
+    right: &Vec<Value>,
+    l_on: &str,
+    r_on: &str,
+    join_type: JoinType,
+    span: Span,
+) -> Value {
+    // Inner / Right Join
+    // ------------------
+    // Make look-up table from rows on left
+    // For each row r on right:
+    //    If any matching rows on left:
+    //        For each matching row l on left:
+    //            Emit (l, r)
+    //    Else if RightJoin:
+    //        Emit (null, r)
+
+    // Left Join
+    // ----------
+    // Make look-up table from rows on right
+    // For each row l on left:
+    //    If any matching rows on right:
+    //        For each matching row r on right:
+    //            Emit (l, r)
+    //    Else:
+    //        Emit (l, null)
+
+    // Outer Join
+    // ----------
+    // Perform Left Join procedure
+    // Perform Right Join procedure, but excluding rows in Inner Join
+
+    let config = Config::default();
+    let sep = ",";
+    let cap = max(left.len(), right.len());
+    let lr_on = if l_on == r_on { Some(l_on) } else { None };
+
+    // For the "other" table, create a map from value in `on` column to a list of the
+    // rows having that value.
+    let mut result: Vec<Value> = Vec::new();
+    let is_outer = matches!(join_type, JoinType::Outer);
+    let (this, other, other_names, join_type) = match join_type {
+        JoinType::Left | JoinType::Outer => (
+            left,
+            lookup_table(right, r_on, sep, cap, &config),
+            column_names(right),
+            // For Outer we do a Left pass and a Right pass; this is the Left
+            // pass.
+            JoinType::Left,
+        ),
+        JoinType::Inner | JoinType::Right => (
+            right,
+            lookup_table(left, l_on, sep, cap, &config),
+            column_names(left),
+            join_type,
+        ),
+    };
+    join_rows(
+        &mut result,
+        this,
+        other,
+        other_names,
+        r_on,
```

Comment:
I'm finding the logic around `l_on`, `r_on`, and `lr_on` difficult to follow. In particular it's surprising that the `this_on` argument here is always `r_on`, but... `this` isn't always `right`.

Is that behaviour correct for left+outer joins (where `this` is `left`)?

Reply:
Thanks @rgwood. I'll take another look at this; I agree the thing you highlighted does seem odd and there could well be a bug. I'll make sure there's test coverage for the things we're worrying about here.

Reply:
Great reviewing @rgwood. I'd completely failed to test different join keys on the left and right, and of course the implementation was not correct as you pointed out. I've added a commit with failing tests of different join keys followed by a commit that fixes it.